### ISSUE TITLE
Excel sheet data not joined properly with .GPKG layer.

### ISSUE THREAD

**Issue by ArnoutAVES** [qgis/QGIS#42971]
Checklist before submitting

- [**√**] Search through existing issue reports and gis.stackexchange.com to check whether the issue already exists
[]((https://docs.qgis.org/testing/en/docs/user_manual/introduction/qgis_configuration.html?highlight=profile#working-with-user-profiles).)
- [**√**] Create a light and self-contained sample dataset and project file which demonstrates the issue
-->
I wasn't able to load the light project in this issue, it says the the file type is not supported.

**Describe the bug**

In QGIS, I have linked data from an excel sheet to a geopackage layer. All the data is showing properly except for these two items, after a lot of trouble shooting I couldn't find out why.
![Capture1](https://user-images.githubusercontent.com/83264404/116187076-76ef3380-a742-11eb-833d-fc8a1d231069.JPG)
![Capture2](https://user-images.githubusercontent.com/83264404/116187083-7bb3e780-a742-11eb-8103-1fc486182f3f.JPG)
![Capture3](https://user-images.githubusercontent.com/83264404/116187084-7bb3e780-a742-11eb-842a-566c69f74223.JPG)

**How to Reproduce**

1. Go to 'ribon'
2. Click on 'open attribute table'
3. Scroll down to 'fid 179 and 182'
4. See error --> the data doesn't match with the data of the excel sheet

**QGIS and OS versions**

QGIS version
3.16.2-Hannover
QGIS code revision
f1660f9da5
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.1.4
Running against GDAL/OGR
3.1.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 10 (10.0)
Active python plugins
db_manager; 
MetaSearch; 
processing

**Additional context**

[AVES Datasheet.xlsx](https://github.com/qgis/QGIS/files/6381971/AVES.Datasheet.xlsx)

**Comment by uclaros** [comment]
Can't really tell without your geopackage... Please check that your geopackage does not already have fields with the same name as the joined fields, or set a proper custom name prefix.

**Comment by roya0045** [comment]
I also suggest checking on the latest  3.16. Just to be safe.

**Comment by ArnoutAVES** [comment]
Dear Stefanos, thank you for looking at my issue.
I have checked it and the geopackage does not already have fields with the
same name as the joined fields, also even after setting a proper custom
name prefix the problem remains.
I have attached the geopackage and the data sheet, hoping that you can
access it now.

**Comment by FelisPimeja** [comment]
@ArnoutAVES , you can attach sample project here if you zip it at first.

**Comment by gioman** [comment]
@ArnoutAVES and to attach files you must post using the GitHub website, you can't answer by email.

**Comment by ArnoutAVES** [comment]
Dear All, Thanks a lot for your support.

I have attached the project, geopackage file and excel data sheet in this message.

[Project.zip](https://github.com/qgis/QGIS/files/6403025/Project.zip)

**Comment by FelisPimeja** [comment]
@ArnoutAVES , something's wrong with your sample dataset.
Source geopackage layer doesn't have attribute "code" 
![image](https://user-images.githubusercontent.com/874261/116665781-d8154200-a9a2-11eb-90d7-143c775252ee.png)

**Comment by ArnoutAVES** [comment]
> I also suggest checking on the latest 3.16. Just to be safe.

I have just tried that and the problem remains. 😊

**Comment by ArnoutAVES** [comment]
> @ArnoutAVES , something's wrong with your sample dataset.
> Source geopackage layer doesn't have attribute "code"
> ![image](https://user-images.githubusercontent.com/874261/116665781-d8154200-a9a2-11eb-90d7-143c775252ee.png)

Thank you for pointing this out.
I have collected an older version of this file from my backup, and added it, it worked fine on the sample project.
However when I added this old file to my main project, it was still not working.
But when I also used the sample project data sheet on my main project, the data got finally linked properly.
I am guessing that it there is some issue with my excel sheet, I will continue to tweak it and see if I can fix this.
Thank you all for your attention and help.
Will post a comment on this thread when I succeed to fix this issue.
